File: parsers/src/main/java/wcl/gpromanager/parsers/utils/ValueRetriever.java
package wcl.gpromanager.parsers.utils;

import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.util.function.Function;

public interface ValueRetriever<T> {

    default T extractValue(Elements elements, String selector){
        return extractValue(elements, selector, s -> s);
    }

    default T extractValue(Document document, String selector) {
        return extractValue(document.children(), selector);
    }

    default T extractValue(Elements elements, String selector, String attributeName){
        return extractValue(elements, selector, attributeName, s -> s);
    }

    default T extractValue(Document document, String selector, String attributeName) {
        return extractValue(document.children(), selector, attributeName);
    }

    default T extractValue(Elements elements, String selector, Function<String, String> transformer) {
        final var value = getValue(elements, selector);
        final var transformedAttributeValue = transform(value, transformer);
        return convert(transformedAttributeValue);
    }

    default T extractValue(Elements elements, String selector, String attributeName, Function<String, String> transformer) {
        final var value = getAttributeValue(elements, selector, attributeName);
        final var transformedAttributeValue = transform(value, transformer);
        return convert(transformedAttributeValue);
    }

    private String getValue(Elements elements, String selector) {
        return elements.select(selector).text();
    }

    private String getAttributeValue(Elements elements, String selector, String attributeName) {
        return elements.select(selector).attr(attributeName);
    }

    private String transform(String value, Function<String, String> transformation) {
        return transformation.apply(value);
    }

    T convert(String attributeValue);
}

File: parsers/src/main/kotlin/wcl/gpromanager/parsers/utils/ValueRetriever.kt
package wcl.gpromanager.parsers.utils

import org.jsoup.nodes.Document
import org.jsoup.select.Elements
import java.util.function.Function

interface ValueRetriever<T> {
    fun extractValue(document: Document, selector: String): T {
        return extractValue(document.children(), selector)
    }

    fun extractValue(document: Document, selector: String, attributeName: String): T {
        return extractValue(document.children(), selector, attributeName)
    }

    fun extractValue(elements: Elements, selector: String, transformer: Function<String, String> = Function { s: String -> s }): T {
        val value = getValue(elements, selector)
        val transformedAttributeValue = transform(value, transformer)
        return convert(transformedAttributeValue)
    }

    fun extractValue(elements: Elements, selector: String, attributeName: String, transformer: Function<String, String> = Function { s: String -> s }): T {
        val value = getAttributeValue(elements, selector, attributeName)
        val transformedAttributeValue = transform(value, transformer)
        return convert(transformedAttributeValue)
    }

    private fun getValue(elements: Elements, selector: String): String {
        return elements.select(selector).text()
    }

    private fun getAttributeValue(elements: Elements, selector: String, attributeName: String): String {
        return elements.select(selector).attr(attributeName)
    }

    private fun transform(value: String, transformation: Function<String, String>): String {
        return transformation.apply(value)
    }

    fun convert(value: String): T
}
